pkg/bpmn/engine.go

```diff
@@ -162,73 +281,101 @@ func (state *Engine) run(instance *processInstanceInfo) (err error) {
 				}
 				commandQueue = append(commandQueue, aCmd)
 			}
-		case activityType:
-			element := cmd.(activityCommand).element
+		case activityCommand:
+			element := tCmd.element
 			originActivity := cmd.(activityCommand).originActivity
-			nextCommands := state.handleElement(ctx, process, instance, element, originActivity)
+			nextCommands, err := engine.handleElement(ctx, batch, process, instance, element, originActivity)
+			if err != nil {
+				return errors.Join(newEngineErrorf("failed to handle activity type element (%+v)", element), err)
+			}
 			commandQueue = append(commandQueue, nextCommands...)
-		case continueActivityType:
-			element := cmd.(continueActivityCommand).activity.Element()
+		case continueActivityCommand:
+			element := tCmd.activity.Element()
 			originActivity := cmd.(continueActivityCommand).originActivity
-			nextCommands := state.handleElement(ctx, process, instance, element, originActivity)
+			nextCommands, err := engine.handleElement(ctx, batch, process, instance, element, originActivity)
+			if err != nil {
+				return errors.Join(newEngineErrorf("failed to handle continue activity type element (%+v)", element), err)
+			}
 			commandQueue = append(commandQueue, nextCommands...)
-		case errorType:
-			err = cmd.(errorCommand).err
-			instance.State = runtime.Failed
-			break
-		case checkExclusiveGatewayDoneType:
-			activity := cmd.(checkExclusiveGatewayDoneCommand).gatewayActivity
-			state.checkExclusiveGatewayDone(activity)
+		case errorCommand:
+			err = tCmd.err
+			// TODO: do something with the error?
+			instance.State = runtime.ActivityStateFailed
+			break mainLoop
```

Comment:
That seems wrong to me.
AFAIK, when an Activity goes Failed, other parallel activities should continue, and breaking the main loop prevents that.